modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/RepositoryStatsCollector.java

```
@@ -0,0 +1,163 @@
+import java.util.Map;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class RepositoryStatsCollector {
+
+    /**
+     * Track billable requests only, required for REST repository metering API
+     */
+    private final EnumMap<StorageOperation, LongAdder> restMetering;
+
+    /**
+     * Telemetry (APM)
+     */
+    private final RepositoriesMetrics telemetry;
+    private final EnumMap<OperationPurpose, EnumMap<StorageOperation, Map<String, Object>>> telemetryAttributes;
+
+    RepositoryStatsCollector(String repoName, RepositoriesMetrics repositoriesMetrics) {
+        this.telemetry = repositoriesMetrics;
+        this.restMetering = new EnumMap<>(StorageOperation.class);
+        for (var op : StorageOperation.values()) {
+            restMetering.put(op, new LongAdder());
+        }
+        this.telemetryAttributes = new EnumMap<>(OperationPurpose.class);
+        for (var purpose : OperationPurpose.values()) {
+            var purposeMap = new EnumMap<StorageOperation, Map<String, Object>>(StorageOperation.class);
+            telemetryAttributes.put(purpose, purposeMap);
+            for (var operation : StorageOperation.values()) {
+                var attrMap = RepositoriesMetrics.createAttributesMap(GoogleCloudStorageRepository.TYPE, repoName, purpose, operation.key);
+            }
+        }
+    }
+
+    /**
+     * Continue collecting metrics with given OperationStats. Useful for readers and writers.
+     */
+    public <T> T continueAndCollect(OperationStats stats, CheckedSupplier<T, IOException> blobFn) throws IOException {
+        OperationStats.set(stats);
+        return blobFn.get();
+    }
+
+    /**
+     * Continue collecting metrics with given OperationStats. Useful for readers and writers.
+     */
+    public void continueAndCollect(OperationStats stats, CheckedRunnable<IOException> runnable) throws IOException {
+        OperationStats.set(stats);
```

I wonder if we should be more assertive about the current state of `OperationStats`. In this case it's probably safe to clear it after each action right? then we could assert that it was empty each time we went to start a new action, to make it fail noisily if they started being interleaved somehow?